tests/components/apns/test_notify.py

```diff
@@ -23,356 +23,366 @@
+    send = mock_client.return_value.send_notification
 
-        assert "5678" == target
-        assert "Hello" == payload.alert
+    yaml_file = {1234: {"name": "test device 1"}}
 
-    @patch("homeassistant.components.apns.notify.APNsClient")
-    @patch("homeassistant.components.apns.notify._write_device")
-    def test_disable_when_unregistered(self, mock_write, mock_client):
-        """Test disabling a device when it is unregistered."""
-        send = mock_client.return_value.send_notification
-        send.side_effect = Unregistered()
+    with patch(
+        "homeassistant.components.apns.notify.load_yaml_config_file",
+        Mock(return_value=yaml_file),
+    ):
+        await _setup_notify(hass)
 
-        yaml_file = {
-            1234: {"name": "test device 1", "tracking_device_id": "tracking123"},
-            5678: {"name": "test device 2", "tracking_device_id": "tracking456"},
-        }
+    assert await hass.services.async_call(
+        "notify",
+        "test_app",
+        {
+            "message": "Hello",
+            "data": {"badge": 1, "sound": "test.mp3", "category": "testing"},
+        },
+        blocking=True,
+    )
 
-        written_devices = []
+    assert send.called
+    assert 1 == len(send.mock_calls)
 
-        def fake_write(_out, device):
-            """Fake write_device."""
-            written_devices.append(device)
+    target = send.mock_calls[0][1][0]
+    payload = send.mock_calls[0][1][1]
 
-        mock_write.side_effect = fake_write
+    assert "1234" == target
+    assert "Hello" == payload.alert
+    assert 1 == payload.badge
+    assert "test.mp3" == payload.sound
+    assert "testing" == payload.category
 
-        with patch(
-            "homeassistant.components.apns.notify.load_yaml_config_file",
-            Mock(return_value=yaml_file),
-        ):
-            self._setup_notify()
 
-        assert self.hass.services.call(
-            "notify", "test_app", {"message": "Hello"}, blocking=True
+@patch("homeassistant.components.apns.notify.APNsClient")
+async def test_send_when_disabled(mock_client, hass):
+    """Test updating an existing device."""
+    send = mock_client.return_value.send_notification
+
+    yaml_file = {1234: {"name": "test device 1", "disabled": True}}
+
+    with patch(
+        "homeassistant.components.apns.notify.load_yaml_config_file",
+        Mock(return_value=yaml_file),
+    ):
+        await _setup_notify(hass)
+
+    assert await hass.services.async_call(
+        "notify",
+        "test_app",
+        {
+            "message": "Hello",
+            "data": {"badge": 1, "sound": "test.mp3", "category": "testing"},
+        },
+        blocking=True,
+    )
+
+    assert not send.called
+
+
+# todo: I had a deadlock issue when making this an async function using the hass fixture
+@patch("homeassistant.components.apns.notify.APNsClient")
+def test_send_with_state(mock_client):
+    """Test updating an existing device."""
```

Comment:
Deadlock happens because initializing `ApnsNotificationService` is calling sync methods. Fixed it

Reply:
Ah, I see, I'll remember `async_add_executor_job` in the future, thanks 👍

Reply:
I found it by running `py.test tests/components/apns/test_notify.py --timeout=.5 -x`. It showed 

```
tests/components/apns/test_notify.py:322: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
homeassistant/components/apns/notify.py:176: in __init__
    track_state_change(hass, tracking_ids, self.device_state_changed_listener)
homeassistant/helpers/event.py:130: in factory
    async_remove = run_callback_threadsafe(
/usr/lib/python3.8/concurrent/futures/_base.py:434: in result
    self._condition.wait(timeout)
```

Calling anything threadsafe from async causes a deadlock 👍